rust/tools/authorization-logic/src/souffle/lowering_ast_datalog.rs

```diff
@@ -0,0 +1,341 @@
+    pub fn lower(prog: &AstProgram) -> DLIRProgram {
+        let mut lowering_pass = LoweringToDatalogPass::new();
+        lowering_pass.prog_to_dlir(prog)
+    }
+
+    fn new() -> LoweringToDatalogPass {
+        LoweringToDatalogPass { fresh_var_count: 0 }
+    }
+
+    fn fresh_var(&mut self) -> String {
+        // Technically it is possible to have a collision with
+        // a programmer-entered variable because the syntax of
+        // identifiers for this language and souffle is the same.
+        // However, the hope is that programmers will not usually
+        // write variable names with 3 underscores.
+        self.fresh_var_count += 1;
+        String::from("x___") + &self.fresh_var_count.to_string()
+    }
+
+    fn verbphrase_to_dlir(&mut self, v: &AstVerbPhrase) -> AstPredicate {
+        match v {
+            AstVerbPhrase::AstPredPhrase { p: pred } => pred.clone(),
+            AstVerbPhrase::AstCanActPhrase { p: prin } => {
+                let mut args_ = Vec::new();
+                args_.push(prin.name.clone());
+                AstPredicate {
+                    name: String::from("canActAs"),
+                    args: args_,
+                }
+            }
+        }
+    }
+
+    fn flat_fact_to_dlir(
+        &mut self,
+        f: &AstFlatFact,
+        speaker: &AstPrincipal,
+    ) -> (AstPredicate, Vec<DLIRAssertion>) {
+        match f {
+            AstFlatFact::AstPrinFact { p, v } => {
+                let v_dlir = self.verbphrase_to_dlir(v);
+                let pred = push_prin(String::from(""), p, &v_dlir);
+
+                // Need to additionally generate:
+                // speaker says x v :- speaker says x canActAs p,
+                //                     speaker says p v
+                // (where x is a fresh principal)
+                let x = AstPrincipal {
+                    name: self.fresh_var(),
+                };
+
+                // speaker says x v;
+                let gen_lhs = push_prin(
+                    String::from("says_"),
+                    speaker,
+                    &push_prin(String::from(""), &x, &v_dlir),
+                );
+                // speaker says x canActAs p
+                let x_as_p = AstPredicate {
+                    name: String::from("canActAs"),
+                    args: [x.name.clone(), p.name.clone()].to_vec(),
+                };
+                let s_says_x_as_p = push_prin(String::from("says_"), speaker, &x_as_p);
+                // speaker says p v
+                let s_says_p_v = push_prin(String::from("says_"), speaker, &pred);
+
+                let gen = DLIRAssertion::DLIRCondAssertion {
+                    lhs: gen_lhs,
+                    rhs: [s_says_x_as_p, s_says_p_v].to_vec(),
+                };
+
+                (pred, [gen].to_vec())
+            }
+            AstFlatFact::AstPredFact { p } => (p.clone(), Vec::new()),
+        }
+    }
+
+    // When reading this, it may be useful to keep in mind that
```

Comment:
nit: move comment inside function. Comments outside of functions should be doc comments explaining what they do.

Reply:
Fixed.